### ISSUE TITLE
Internal Failure while saving the schema: GraphQLAPITransformerSchema - UPDATE_ROLLBACK_FAILED

### ISSUE THREAD

**Issue by justinjaeger** [aws-amplify/amplify-category-api#1040]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### How did you install the Amplify CLI?

yarn

### If applicable, what version of Node.js are you using?

v16.15.1

### Amplify CLI Version

10.0.0

### What operating system are you using?

Mac

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

I went into CloudFormation and fixed the Drift States

### Amplify Categories

api

### Amplify Commands

push

### Describe the bug

I performed an amplify push which:
- removes an enum that is being used by a field
- removes the field using the enum (this field was marked as mandatory, maybe that's part of the issue)
- adds a new enum
- adds a new table

It resulted in the following error message:

```
The following resource(s) failed to update: [GraphQLAPITransformerSchema3CB2AE18]

UPDATE_FAILED GraphQLAPITransformerSchema3CB2AE18 AWS::AppSync::GraphQLSchema Mon Nov 28 2022 22:58:23 GMT-0500 (Eastern Standard Time) Schema Creation Status is FAILED with details: Internal Failure while saving the schema.

```

Then it attempts to roll back and fails. Because the rollback fails, I'm unable to push any new changes.

I went into CloudFormation and resolved the inconsistent drift states, then did "continue update rollback" with no success. I checked the CloudFormation events on the failed stack, and the first message is this:

`Schema Creation Status is FAILED with details: Internal Failure while saving the schema.`

So it appears something is wrong with my GraphQL schema which is preventing the rollback from continuing.

I went into AppSyncConsole to see if the schema had any attributes from the NEW schema (the one that failed), and it didn't have any obvious differences. Perhaps it would be fixed if I could somehow re-generate this file from the schema.graphql in my code (the one that worked previously), but I'm not sure how to do that.

### Expected behavior

`amplify push` was expected to either modify my graphql schema via CloudFormation stack, or roll it back safely, but the update failed when saving the schema.

### Reproduction steps

I've attached two files, schema_before.graphql and schema_after.graphql.

schema_before.graphql worked, and the update to schema_after.graphql is led to failure.
[schema_after.graphql.txt](https://github.com/aws-amplify/amplify-cli/files/10110174/schema_after.graphql.txt)
[schema_before.graphql.txt](https://github.com/aws-amplify/amplify-cli/files/10110175/schema_before.graphql.txt)

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line

# I'm pasting the schema before, then the schema after:

# SCHEMA BEFORE:
## This "input" configures a global authorization rule to enable public access to
## all models in this schema. Learn more about authorization rules here: https://docs.amplify.aws/cli/graphql/authorization-rules
# input AMPLIFY { globalAuthRule: AuthRule = { allow: public } } # FOR TESTING ONLY!

enum UserRole {
  ADMIN
  USER
}

enum AwardsBody {
  ACADEMY_AWARDS
  GOLDEN_GLOBES
  CRITICS_CHOICE
  BAFTA
  HCA
  PGA
  SAG
  DGA
  WGA
  ADG
  MAKEUP_GUILD
  CDG
  ASC
  MPSE
}

enum EventType {
    WIN
    NOMINATION
}

enum CategoryType {
    FILM
    PERFORMANCE
    SONG
}

enum CategoryName {
    PICTURE
    DIRECTOR
    ACTOR
    ACTRESS
    SUPPORTING_ACTOR
    SUPPORTING_ACTRESS
    ORIGINAL_SCREENPLAY
    ADAPTED_SCREENPLAY
    SCREENPLAY
    INTERNATIONAL
    ANIMATED
    DOCUMENTARY
    EDITING
    CINEMATOGRAPHY
    PRODUCTION_DESIGN
    COSTUMES
    MAKEUP
    VISUAL_EFFECTS
    SOUND
    SCORE
    SONG
    SHORT_ANIMATED
    SHORT_DOCUMENTARY
    SHORT_LIVE_ACTION
    ENSEMBLE
    COMEDY_PICTURE
    COMEDY_ACTOR
    COMEDY_ACTRESS
    ACTION_PICTURE
    SCIFI_HORROR_PICTURE
    YOUNG_ACTOR
    RISING_STAR
    DEBUT
    FIRST_SCREENPLAY
    BRITISH_PICTURE
    ANIMATED_PERFORMANCE
    BLOCKBUSTER
    ACTING_ACHIEVEMENT
    FEMALE_DIRECTOR
    MALE_DIRECTOR
    INDIE_PICTURE
    BREAKTHROUGH
}

type User @model @auth(rules: [{allow: public}]) {
    id: ID! @primaryKey
    email: String!
    username: String
    name: String
    bio: String
    image: String
    role: UserRole! @default(value: "USER")}

type PredictionSet @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    user: User! @hasOne
    event: Event! @hasOne
    category: Category! @hasOne
    predictions: [Prediction] @hasMany
}

type Prediction @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    user: User! @hasOne
    contender: Contender! @belongsTo
    ranking: Int!
}

type Event @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    categories: [Category] @hasMany # Retrieve categories on an event record
    awardsBody: AwardsBody!
    year: Int!
    type: EventType!
    expiration: AWSDateTime
    isActive: String @default(value: "x") # essentially boolean
    contenders: [Contender] @hasMany
}

type Category @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    name: CategoryName!
    type: CategoryType! @default(value: "FILM")
    event: Event! @belongsTo # Lets you retrieve info about the event from a Category record
    contenders: [Contender] @hasMany
}

# Belongs to a particular category
type Contender @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    category: Category! @belongsTo # belongsTo is for bi-directional relationships. Category also hasMany contenders
    event: Event! @belongsTo
    movie: Movie! @hasOne
    person: Person @hasOne # optional - only applies to acting categories
    song: Song @hasOne # optional - only applies to the song category
    predictions: [Prediction] @hasMany
    didReceiveNominationOrWin: Boolean
}

type Movie @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    tmdbId: Int! # unique value
    studio: String # distribution studio
}

type Person @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    tmdbId: Int! # unique value
}
  
type Song @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    title: String!
    artist: String!
    movie: Movie! @hasOne
}

```

```graphql
# Put schemas below this line

# SCHEMA_AFTER

## This "input" configures a global authorization rule to enable public access to
## all models in this schema. Learn more about authorization rules here: https://docs.amplify.aws/cli/graphql/authorization-rules
# input AMPLIFY { globalAuthRule: AuthRule = { allow: public } } # FOR TESTING ONLY!

enum UserRole {
  ADMIN
  USER
}

enum AwardsBody {
  ACADEMY_AWARDS
  GOLDEN_GLOBES
  CRITICS_CHOICE
  BAFTA
  HCA
  PGA
  SAG
  DGA
  WGA
  ADG
  MAKEUP_GUILD
  CDG
  ASC
  MPSE
}

enum Boolean {
    TRUE
    FALSE
}

enum CategoryType {
    FILM
    PERFORMANCE
    SONG
}

enum CategoryName {
    PICTURE
    DIRECTOR
    ACTOR
    ACTRESS
    SUPPORTING_ACTOR
    SUPPORTING_ACTRESS
    ORIGINAL_SCREENPLAY
    ADAPTED_SCREENPLAY
    SCREENPLAY
    INTERNATIONAL
    ANIMATED
    DOCUMENTARY
    EDITING
    CINEMATOGRAPHY
    PRODUCTION_DESIGN
    COSTUMES
    MAKEUP
    VISUAL_EFFECTS
    SOUND
    SCORE
    SONG
    SHORT_ANIMATED
    SHORT_DOCUMENTARY
    SHORT_LIVE_ACTION
    ENSEMBLE
    COMEDY_PICTURE
    COMEDY_ACTOR
    COMEDY_ACTRESS
    ACTION_PICTURE
    SCIFI_HORROR_PICTURE
    YOUNG_ACTOR
    RISING_STAR
    DEBUT
    FIRST_SCREENPLAY
    BRITISH_PICTURE
    ANIMATED_PERFORMANCE
    BLOCKBUSTER
    ACTING_ACHIEVEMENT
    FEMALE_DIRECTOR
    MALE_DIRECTOR
    INDIE_PICTURE
    BREAKTHROUGH
}

type User @model @auth(rules: [{allow: public}]) {
    id: ID! @primaryKey
    email: String!
    username: String
    name: String
    bio: String
    image: String
    role: UserRole! @default(value: "USER")}

# CURRENT predictions only
type PredictionSet @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    user: User! @hasOne
    event: Event! @hasOne
    category: Category! @hasOne
    predictions: [Prediction] @hasMany
}

type HistoryPredictionSet @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    user: User! @hasOne
    event: Event! @hasOne
    category: Category! @hasOne
    predictions: [Prediction] @hasMany
}

type Prediction @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    user: User! @hasOne
    contender: Contender! @belongsTo
    ranking: Int!
}

type Event @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    categories: [Category] @hasMany # Retrieve categories on an event record
    awardsBody: AwardsBody!
    year: Int!
    contenders: [Contender] @hasMany
    nominationsOpen: Boolean! @default(value: "FALSE")
    winsOpen: Boolean! @default(value: "FALSE")
    nominationDateTime: AWSDateTime # nullable. must have ability to set as admin
    winDateTime: AWSDateTime # nullable. must have ability to set as admin
}

type Category @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    name: CategoryName!
    type: CategoryType! @default(value: "FILM")
    event: Event! @belongsTo # Lets you retrieve info about the event from a Category record
    contenders: [Contender] @hasMany
}

# Belongs to a particular category
type Contender @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    category: Category! @belongsTo # belongsTo is for bi-directional relationships. Category also hasMany contenders
    event: Event! @belongsTo
    movie: Movie! @hasOne
    person: Person @hasOne # optional - only applies to acting categories
    song: Song @hasOne # optional - only applies to the song category
    predictions: [Prediction] @hasMany
}

type Movie @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    tmdbId: Int! # unique value
    studio: String # distribution studio
}

type Person @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    tmdbId: Int! # unique value
}
  
type Song @model @auth(rules: [{ allow: public }]) {
    id: ID! @primaryKey
    title: String!
    artist: String!
    movie: Movie! @hasOne
}



```
</details>

### Project Identifier

[report-1669701544046.zip](https://github.com/aws-amplify/amplify-cli/files/10110180/report-1669701544046.zip)

### Log output

<details>

```
# Put your logs below this line

UPDATE_FAILED GraphQLAPITransformerSchema3CB2AE18 AWS::AppSync::GraphQLSchema Mon Nov 28 2022 22:58:23 GMT-0500 (Eastern Standard Time) Schema Creation Status is FAILED with details: Internal Failure while saving the schema.

UPDATE_FAILED      GraphQLAPITransformerSchema3CB2AE18 AWS::AppSync::GraphQLSchema Mon Nov 28 2022 22:59:08 GMT-0500 (Eastern Standard Time) 
***Resource update cancelled                                                                                                                                                                            

UPDATE_FAILED      GraphQLAPIDefaultApiKey215A6DD7     AWS::AppSync::ApiKey        Mon Nov 28 2022 22:59:07 GMT-0500 (Eastern Standard Time) API key must be valid for a minimum of 1 days. (Service: AWSAppSync; Status Code: 400; Error Code: ApiKeyValidityOutOfBoundsException; Request ID: dddddce7-9554-4cb3-aba2-3b4ad99993d9; Proxy: null)

⠋ Rolling back (2 of 2)🛑 An error occurred during the push operation: /
["Index: 1 State: {\"deploy\":\"waitingForDeployment\"} Message: Resource is not in the state stackUpdateComplete","Index: 0 State: {\"rollback\":\"rollingBack\"} Message: Stack:arn:aws:cloudformation:us-east-1:709934184485:stack/amplify-awardsapp-dev-115717-apiawardsapp-YRXGG09GONN3/7afbf920-4ca2-11ed-9a4e-12460186cb85 is in UPDATE_ROLLBACK_FAILED state and can not be updated."]
```

</details>

**Comment by josefaidt** [comment]
Hey @justinjaeger :wave: thanks for raising this! Although I've experienced the "internal server error" issue when pushing a new/updated schema, I did notice in the log output there's another error that stands out and could potentially be the cause here.

>  API key must be valid for a minimum of 1 days.

Can you check to see whether the API key has expired or the validity shows a date in the past (in the API's CloudFormation template)

**Comment by justinjaeger** [comment]
> Hey @justinjaeger 👋 thanks for raising this! Although I've experienced the "internal server error" issue when pushing a new/updated schema, I did notice in the log output there's another error that stands out and could potentially be the cause here.
> 
> > API key must be valid for a minimum of 1 days.
> 
> Can you check to see whether the API key has expired or the validity shows a date in the past (in the API's CloudFormation template)

Hi Josef, thanks for getting back to me!

Looking at my CloudFormation template, I found this:

        "GraphQLAPIDefaultApiKey215A6DD7": {
            "Type": "AWS::AppSync::ApiKey",
            "Properties": {
                "ApiId": {
                    "Fn::GetAtt": [
                        "GraphQLAPI",
                        "ApiId"
                    ]
                },
                "Expires": 1668070165
            }
        },

Not sure how to decode the expiration. It's a datatype "double", and is "[represented as seconds since the epoch, rounded down to the nearest hour](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-appsync-apikey.html#cfn-appsync-apikey-expires)."

**Comment by justinjaeger** [comment]
Hey @josefaidt would appreciate a follow up when you get a change, thank you!

**Comment by josefaidt** [comment]
Hey @justinjaeger thanks for clarifying! Looking at the expiration epoch it appears it expired Nov 10 (`Thursday, November 10, 2022 8:49:25 AM`), however since you mentioned you do not see any expired keys in the AppSync console is it possible this was rotated manually? 

Additionally, in the CloudFormation Console it is possible to continue the rollback and bypass the checks for the API key. This should at least allow you to continue, and if you run into additional issues you can try setting `CreateApiKey` to 0, push, then set it back to 1 to regenerate the epoch

**Comment by josefaidt** [comment]
related https://github.com/aws-amplify/amplify-category-api/issues/636

If the additional note above is also unsuccessful, we can try using [overrides](https://docs.amplify.aws/cli/graphql/override/#customize-amplify-generated-appsync-graphql-api-resources) to adjust the API key expiry and sync your templates

**Comment by Shawkat-Roy** [comment]
>  in the CloudFormation Console it is possible to continue the rollback and bypass the checks for the API key.

Hi @josefaidt I hope you don't mind a quick question about the successful solution you provided.

I'm facing the same issue that was successfully solved here.
My stack is in the state of failed update rollback, and it's because the API key it's trying to roll back to is outdated. You mentioned that it's possible to bypass the checks for the API key, and this seemed to work successfully for Justin.

Can you explain to me how in the CloudFormation Console I can rerun the rollback while bypassing the API key check?
Or perhaps can you point me to some documentation on how to do this? I can't seem to figure it out.

I found documentation on how to skip rolling back the resource, using AWS CLI, but it seems that's slightly different to what you proposed, or am I wrong, is that what you were actually suggesting?

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-continueupdaterollback.html

Thanks for your time.

**Comment by squirrelhomie** [comment]
I'm having a similar issue. My stack is in UPDATE_ROLLBACK_COMPLETE and when I run amplify push I get this error - DeploymentFault: Resource is not in the state stackUpdateComplete

At this point, I can't remember what I did exactly. I believe the api key was expired and at some point, deleted the api key in app sync, created a new one.

I've tried CreateAPIKey: 0 with no helpful errors following. When I add APIKeyExpirationEpoch: -1 to my parameters.json I get a somewhat more helpful error:

Resource Name: GraphQLAPIDefaultApiKey<key> (AWS::AppSync::ApiKey)
Event Type: update
Reason: API key not found: <api-key> (Service: AWSAppSync; Status Code: 404; Error Code: NotFoundException; Request ID: ; Proxy: null)
The api key it is referencing is one that is found in my aws-exports but I've since deleted that key from app sync.

At this point, not sure what direction to go. Any help would be appreciated!

**Comment by Sushil787** [comment]
@justinjaeger  I fell in to same issue 
The following resource(s) failed to update: [GraphQLAPITransformerSchema3CB2AE18, GraphQLAPIDefaultApiKey215A6DD7].

API key not found: da2-ramya5pxtbcgjc4om5dbdy66ny (Service: AWSAppSync; Status Code: 404; Error Code: NotFoundException; Request ID: 7df836e7-a60a-426a-aca6-dd3024388029; Proxy: null)


-
Resource update cancelled
This is my cloud front template. 
I am trying to update the expires time for key but it's not working. 
I have deleted api key from app sync and created new but no changes are reflecting in cloud front how can i update key and expiration time. I have changed key from local and pushed but no success. 

    "GraphQLAPIDefaultApiKey215A6DD7": {
            "Type": "AWS::AppSync::ApiKey",
            "Properties": {
                "ApiId": {
                    "Fn::GetAtt": [
                        "GraphQLAPI",
                        "ApiId"
                    ]
                },
                "Expires": 1732380386
            }
        },

This is my Session Identifier: 68722762-efcd-4f5f-891a-4877c8f02d87
